hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java

```diff
@@ -360,20 +360,33 @@ private void watchForCommit(boolean bufferFull) throws IOException {
   }
 
   private CompletableFuture<ContainerProtos.
```

Comment:
What is the use of force flag here? can you update the comments here as well?

Reply:
The `force` flag is needed to handle the case when the stream is closed right after a putBlock, without further new data.  Since we don't know in advance which putBlock will be the last, we need to "force" another putBlock in this case.

Reply:
> The `force` flag is needed to handle the case when the stream is closed right after a putBlock, without further new data. Since we don't know in advance which putBlock will be the last, we need to "force" another putBlock in this case.

If my understanding , if a stream gets closed, as a part of close itself it should send the last putBlock. Also, we always know which putBlock will be last, with this putBlock as cumulative length of data written will be higher and same putBlock future is tracked in futureMap in commitWatcher.

Reply:
> if a stream gets closed, as a part of close itself it should send the last putBlock

That's happening here.

> we always know which putBlock will be last

This is a stream, caller can write as much data as wanted (within block size limit).  How do we know which will be last before `close()` is called?

Reply:
> > if a stream gets closed, as a part of close itself it should send the last putBlock
> 
> That's happening here.
> 
> > we always know which putBlock will be last
> 
> This is a stream, caller can write as much data as wanted (within block size limit). How do we know which will be last before `close()` is called?

What I meant here was a part of close, whatever putBlock request sent will always be the last. But i get why force flag is being used here. Its the case where, for example, let's we did a flush first and then did a close without writing anything in between. In such case, even though no data has been written in between, a putBlock needs to be enforced. Am i correct?

Reply:
> Am i correct?

Yes.